### ISSUE TITLE
Style/NumericPredicate gets overridden by Style/NumericLiterals

### ISSUE THREAD

**Issue by jdashton** [rubocop/rubocop#10337]
Rubocop almost always does an amazing job of detecting and correcting multiple offenses in a single line of code. However, in this case one correction overwrote the other correction, leaving an offense waiting for a subsequent run. This may have been exacerbated by another of the bugs I reported today (on Style/TernaryParentheses). When that offense is disabled, the `Style/NumericPredicate` cop shows up twice and counts as two offenses, even though it is fixing the same offense both times.

```
lib/aoc2021/puzzles/amphipod.rb:41:55: C: [Corrected] Style/NumericPredicate: Use (count % 10000).zero? instead of count % 10000 == 0. (https://rubystyle.guide#predicate-methods)
          puts "#{ count }...#{ next_move.score }" if count % 10000 == 0
                                                      ^^^^^^^^^^^^^^^^^^
lib/aoc2021/puzzles/amphipod.rb:41:55: C: [Corrected] Style/NumericPredicate: Use (count % 10_000).zero? instead of count % 10_000 == 0. (https://rubystyle.guide#predicate-methods)
          puts "#{ count }...#{ next_move.score }" if count % 10_000 == 0
                                                      ^^^^^^^^^^^^^^^^^^^
lib/aoc2021/puzzles/amphipod.rb:41:63: C: [Corrected] Style/NumericLiterals: Use underscores(_) as thousands separator and separate every 3 digits with them. (https://rubystyle.guide#underscores-in-numerics)
          puts "#{ count }...#{ next_move.score }" if count % 10000 == 0
                                                              ^^^^^
```

--------

## Expected behavior

Cops should correct precisely the observed offense, doing it in such a way that other offenses on the same line can also be corrected during the same run.

## Actual behavior

Both the Style/NumericPredicate and Style/NumericLiterals cops reported that they had performed corrections. However, the resulting code did not show the correction from Style/NumericPredicate.

## Steps to reproduce the problem

Input:
```
          puts "#{ count }...#{ next_move.score }" if count % 10000 == 0
```
Execution:
```
$ bin/rubocop --debug -A lib/aoc2021/puzzles/amphipod.rb
For /Users/jdashton/cSens/learn/aoc2021-ruby: configuration from /Users/jdashton/cSens/learn/aoc2021-ruby/.rubocop.yml
Default configuration from /Users/jdashton/.rvm/gems/ruby-3.1.0@aoc2021/gems/rubocop-1.24.0/config/default.yml
Inspecting 1 file
Scanning /Users/jdashton/cSens/learn/aoc2021-ruby/lib/aoc2021/puzzles/amphipod.rb
E

Offenses:

... erroneous Lint/Syntax offenses elided...

lib/aoc2021/puzzles/amphipod.rb:41:55: C: [Corrected] Style/NumericPredicate: Use (count % 10000).zero? instead of count % 10000 == 0. (https://rubystyle.guide#predicate-methods)
          puts "#{ count }...#{ next_move.score }" if count % 10000 == 0
                                                      ^^^^^^^^^^^^^^^^^^
lib/aoc2021/puzzles/amphipod.rb:41:63: C: [Corrected] Style/NumericLiterals: Use underscores(_) as thousands separator and separate every 3 digits with them. (https://rubystyle.guide#underscores-in-numerics)
          puts "#{ count }...#{ next_move.score }" if count % 10000 == 0
                                                              ^^^^^
lib/aoc2021/puzzles/amphipod.rb:212:34: C: [Corrected] Style/TernaryParentheses: Omit parentheses for ternary conditions.
          @board[2][room_spot] = (md[1].to_sym in AMPHIPODS) ? md[1].to_sym : :empty
                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

1 file inspected, 13 offenses detected, 3 offenses corrected
Finished in 0.8476939999964088 seconds
```
Result:
```
          puts "#{ count }...#{ next_move.score }" if count % 10_000 == 0
```

## RuboCop version

```
$ bin/rubocop -V
1.24.0 (using Parser 3.0.3.2, rubocop-ast 1.15.1, running on ruby 3.1.0 x86_64-darwin21)
```

**Comment by dvandersluis** [comment]
@jdashton can you confirm this is still happening for you on the latest version of RuboCop?

When I run `rubocop test.rb -A --only Style/NumericPredicate,Style/NumericLiterals` with this code:
```ruby
puts "#{ count }...#{ next_move.score }" if count % 10000 == 0
```

it gets corrected to
```ruby
puts "#{ count }...#{ next_move.score }" if (count % 10_000).zero?
```

**Comment by jdashton** [comment]
Yes, I confirm that this is still happening on Rubocop version 1.26.1:
```
$ bin/rubocop --version
1.26.1
```
The corrections occur, as you assert above. However, as I documented in the original problem statement, the NumericPredicate correction happens twice, like so:
```
$ bin/rubocop lib/aoc2021/puzzles/amphipod.rb -A --only Style/NumericPredicate,Style/NumericLiterals
Inspecting 1 file
C

Offenses:

lib/aoc2021/puzzles/amphipod.rb:37:50: C: [Corrected] Style/NumericPredicate: Use (count % 10000).zero? instead of (count % 10000) == 0. (https://rubystyle.guide#predicate-methods)
          puts "#{ count }...#{ move.score }" if (count % 10000) == 0
                                                 ^^^^^^^^^^^^^^^^^^^^
lib/aoc2021/puzzles/amphipod.rb:37:50: C: [Corrected] Style/NumericPredicate: Use (count % 10_000).zero? instead of (count % 10_000) == 0. (https://rubystyle.guide#predicate-methods)
          puts "#{ count }...#{ move.score }" if (count % 10_000) == 0
                                                 ^^^^^^^^^^^^^^^^^^^^^
lib/aoc2021/puzzles/amphipod.rb:37:59: C: [Corrected] Style/NumericLiterals: Use underscores(_) as thousands separator and separate every 3 digits with them. (https://rubystyle.guide#underscores-in-numerics)
          puts "#{ count }...#{ move.score }" if (count % 10000) == 0
                                                          ^^^^^

1 file inspected, 3 offenses detected, 3 offenses corrected
```
Each of those corrections seems to count as a single offense, and so Rubocop reports `3 offenses corrected` when, by human standards, only 2 offenses have actually been corrected.

**Comment by jdashton** [comment]
I should amend my confirmation: both types of errors are now being corrected, as you asserted on March 26. That is different from my original problem statement. However, the NumericPredicate cop is still generating two lines of output and counting for two offenses, and the transformation of ` == 0` to `.zero?` apparently happens in duplicate.

**Comment by dvandersluis** [comment]
RuboCop does multiple passes on code that is autocorrected, and it's possible that the autocorrection for one is undoing the autocorrection for the other in a transient state. I wouldn't worry too much about the absolute count of offenses to be honest, there are many ways for a single offense to cascade into a ton of offenses due to the autocorrection loop (for instance, a piece of code is replaced with another piece of code and then the layout has to be reflowed).

I am going to close this issue since there is no bug anymore.
